source/solvers/navier_stokes_base.cc

```diff
@@ -900,27 +900,83 @@ NavierStokesBase<dim, VectorType, DofsType>::refine_mesh_kelly()
-            present_solution,
-            estimated_error_per_cell,
-            this->fe->component_mask(pressure));
+          if constexpr (std::is_same_v<VectorType,
+                                       TrilinosWrappers::MPI::Vector> ||
+                        std::is_same_v<VectorType,
+                                       TrilinosWrappers::MPI::BlockVector>)
+            {
+              KellyErrorEstimator<dim>::estimate(
+                *this->mapping,
+                this->dof_handler,
+                *this->face_quadrature,
+                typename std::map<types::boundary_id,
+                                  const Function<dim, double> *>(),
+                present_solution,
+                estimated_error_per_cell,
+                this->fe->component_mask(pressure));
+            }
+          if constexpr (std::is_same_v<
+                          VectorType,
+                          LinearAlgebra::distributed::Vector<double>>)
+            {
+              VectorType locally_relevant_solution;
+              locally_relevant_solution.reinit(this->locally_owned_dofs,
+                                               this->locally_relevant_dofs,
+                                               this->mpi_communicator);
+              locally_relevant_solution.copy_locally_owned_data_from(
+                this->present_solution);
+              locally_relevant_solution.update_ghost_values();
+
+              KellyErrorEstimator<dim>::estimate(
+                *this->mapping,
+                this->dof_handler,
+                *this->face_quadrature,
+                typename std::map<types::boundary_id,
+                                  const Function<dim, double> *>(),
+                locally_relevant_solution,
+                estimated_error_per_cell,
+                this->fe->component_mask(pressure));
+            }
         }
       else if (ivar.first == Parameters::MeshAdaptation::Variable::velocity)
         {
-          KellyErrorEstimator<dim>::estimate(
-            *this->mapping,
-            this->dof_handler,
-            *this->face_quadrature,
-            typename std::map<types::boundary_id,
-                              const Function<dim, double> *>(),
-            present_solution,
-            estimated_error_per_cell,
-            this->fe->component_mask(velocity));
+          if constexpr (std::is_same_v<VectorType,
+                                       TrilinosWrappers::MPI::Vector> ||
+                        std::is_same_v<VectorType,
+                                       TrilinosWrappers::MPI::BlockVector>)
+            {
+              KellyErrorEstimator<dim>::estimate(
+                *this->mapping,
+                this->dof_handler,
+                *this->face_quadrature,
+                typename std::map<types::boundary_id,
+                                  const Function<dim, double> *>(),
+                present_solution,
```

Comment:
What type of ghosting does `present_solution` have?

Reply:
It is declared here:
https://github.com/lethe-cfd/lethe/blob/6dc4461ccfc72a4e2ba0ec572e467197a7614544/include/solvers/navier_stokes_base.h#L441

Initialized for the `lethe-fluid-matrix-free` application using the matrix-free operator here: 
https://github.com/lethe-cfd/lethe/blob/6dc4461ccfc72a4e2ba0ec572e467197a7614544/source/solvers/mf_navier_stokes.cc#L209

Reply:
Thanks! 

May I propose an other solution to the problem. Always copy the values from `present_solution` to a locally relevant version (AMR is not a performance-critical part of the code). This is done at other places in lethe.

Reply:
I don't think you need to use `copy_locally_owned_data_from()` but you can simply use the copy assignment operator.